include/unifex/new_thread_context.hpp

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2020-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/get_stop_token.hpp>
+#include <unifex/receiver_concepts.hpp>
+
+#include <cassert>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace unifex {
+
+
+class new_thread_context {
+private:
+  template<typename Receiver>
+  class schedule_op {
+  public:
+    template<typename Receiver2>
+    explicit schedule_op(new_thread_context* ctx, Receiver2&& r)
+    : ctx_(ctx), receiver_((Receiver2&&)r)
+    {
+    }
+
+    ~schedule_op() {
+        assert(!thread_.joinable());
+    }
+
+    void start() & noexcept {
+      {
+        std::unique_lock lk{ctx_->mut_};
+        ++ctx_->activeThreadCount_;
+      }
+
+      try {
+        std::lock_guard opLock{mut_};
+        thread_ = std::thread([this]() noexcept { this->run(); });
+      } catch (...) {
+        {
+          std::lock_guard ctxLock{ctx_->mut_};
+          --ctx_->activeThreadCount_;
+          // Don't need to signal the condition_variable here as the
+          // operation has not yet completed (we haven't called set_error)
+          // and so the caller shouldn't be calling the destructor of the
+          // new_thread_context yet.
+        }
+        unifex::set_error(std::move(receiver_), std::current_exception());
+      }
+    }
+
+  private:
+    void run() noexcept {
+      std::thread thisThread;
+      {
+        std::lock_guard opLock{mut_};
+        thisThread = std::move(thread_);
```

Comment:
The assumption here is that `thisThread` ends up referring to the thread created on line 52, but I don't think that's necessarily the case. It's possible this code executed *before* the thread created on line 52 is assigned to `thread_`. Also, isn't it possible for another thread object to get written to `thread_` in a racy way?

Reply:
The `thread_` variable is a member of the operation-state class and so should have a 1:1 correspondence with a particular thread-creation. This variable is assigned only by the `start()` method that created the thread and is only read by the newly created thread.

We create the `std::thread()` while holding the `mut_` lock, so the new thread will not be able to acquire the lock until the thread that created it releases the lock, which only happens after we have completed assigning to the `thread_` variable.

Reply:
@kirkshoop also had the same concern on first reading.
I'll add a comment here to explain as I'm sure others will probably have similar concerns.

Reply:
Ah. I got mixed up and thought this was the std::thread in the context.